examples/webgl_loader_gainmap.html

```diff
@@ -0,0 +1,352 @@
+<!DOCTYPE html>
+<html lang="en">
+	<head>
+		<title>three.js webgl - gainmap hdr</title>
+		<meta charset="utf-8">
+		<meta name="viewport" content="width=device-width, user-scalable=no, minimum-scale=1.0, maximum-scale=1.0">
+		<link type="text/css" rel="stylesheet" href="main.css">
+		<style>
+			.lbl {
+				color: #fff;
+				font-size: 16px;
+				font-weight: bold;
+				position: absolute;
+				bottom: 0px;
+				z-index: 100;
+				text-shadow: #000 1px 1px 1px;
+				background-color: rgba(0,0,0,0.85);
+				padding: 1em;
+			}
+
+			#lbl_left {
+				text-align:left;
+				left:0px;
+			}
+		</style>
+	</head>
+
+	<body>
+		<div id="info">
+			<a href="https://threejs.org" target="_blank" rel="noopener">three.js</a> - gain map (ultra hdr) loader <br/>
+			Gain map images converted from hdr with <a href="https://gainmap-creator.mono-grid.com/" target="_blank" rel="noopener">Gain map converter</a>. <br />
+			See external <a href="https://github.com/MONOGRID/gainmap-js" target="_blank" rel="noopener">gainmap-js</a> for more information on how to use and create gain map images.
+		</div>
+
+		<div id="lbl_left" class="lbl"></div>
+
+		<script type="importmap">
+			{
+				"imports": {
+					"three": "../build/three.module.js",
+					"three/addons/": "./jsm/",
+					"@monogrid/gainmap-js": "https://unpkg.com/@monogrid/gainmap-js@2.0.0/dist/decode.js"
```

Comment:
Should we check in the JPEGRLoader to `three/addons/loaders`? I'm not sure how @mrdoob feels about external references.

Reply:
it has been done before [here for example](https://github.com/mrdoob/three.js/blob/master/examples/webgl_raycaster_bvh.html).

my idea is to start small with an "external" example and see if there's enough interest in the technology to justify a three.js "official" loader.

A "proper" integration would require the addition of the whole [decode](https://github.com/MONOGRID/gainmap-js/tree/main/src/decode) part of the code which includes shaders, materials, etc.

Plus our loaders are a bit "atypical": 
1. they require a `WebGLRenderer` to work: decoding is done in the GPU because it is doable and is order of magnitude faster than doing it in JS. Our first implementation was done in JS code but it had to loop the whole HDR Uint16Array and it was too slow for my tastes :) 
2. they return [an object](https://github.com/MONOGRID/gainmap-js/blob/main/src/core/QuadRenderer.ts) which contain
  a. a renderTarget where the texture is rendered, usable elsewhere with `renderTarget.texture`
  b. methods to obtain a `DataTexture` from the GPU, [if needed](https://github.com/mrdoob/three.js/issues/27171#issuecomment-1806428518)

I'm not sure if three.js's itself wants to break established Loader *conventions* with this, for the moment we can keep these "external" and see if it works for users. 

Let me know otherwise!

Reply:
> Our first implementation was done in JS code but it had to loop the whole HDR Uint16Array and it was too slow for my tastes :)

How slow was that? 🤔

JS parsers tend to be much easier to maintain and improve by the community.
Plus can be used in environments without GPUs.

Reply:
Hi! Nice to speak to you, me and my company are **avid** Three.js users since the beginning, I'd like to take the opportunity to thank you all for the hard work of maintaining the best WebGL Library :rocket: 

> How slow was that? 🤔

it was more or less the same speed of the EXR or RGBM parsers you already have, all the decoder does is loop the whole `sdr_image(x, y)` and do simple math operations on the RGB values.

In my opinion, though, the JS decoder implementation kinda defeated the purpose of the library which is useful because gain maps finally allows scenes to have 4k or 8k HDR Environments (or high res HDR light maps) with decent file sizes without forcing the users to wait a long time. With 8k images the JS decoder parsed the image in **more time** than it took to download it and we felt it kinda ruined the fun.

> JS parsers tend to be much easier to maintain and improve by the community.
> Plus can be used in environments without GPUs.

I agree, we used to do the decoding on a worker before implementing the GPU solution.

I can bring back the JS code if you prefer but I'm not very fond of the results, I feel it could be an alternative to the GPU decoder but the default (and the loaders) would still be GPU based.

Reply:
TBH, I would also prefer a JS solution for the reasons @mrdoob mentioned. 

Besides, it seems `decode.js` relies on WebGL for parsing the HDR data which isn't so greta in context of WebGPU.

Reply:
I wrote that `PMREMGenerator`, so I'm glad it's not my bug 😄 

Maybe it's just because of the custom shaders I wrote for the above, but I tend to agree that a small quad shader seems like a good approach here, assuming we quash the NaNs.

Reply:
@Mugen87 is there a better approach to a custom shader that would more easily translate between WebGL & WebGPU?

Reply:
> Could you share a URL to an example that uses the JS decoder so we can check the parsing times?

https://monogrid.github.io/gainmap-js/jstest.html

this is a bit quick & dirty but it gets the idea through, the code I pasted in the source is taken directly from our old JS implementation in its naive-ish form possible. It gets the job done and you get an `Uint16Array` back from two `Uint8Array` s, it is basically version 0.0.1 of the whole library :smile: 

I get this result on my machine: 
```
Loading 4k images
Download 4k images time: 736ms
CPU Decode time for 4k image: 2462ms
Loading 8k images
Download 8k images time: 1640ms
CPU Decode time for 4k image: 9782ms
```

> Maybe it's just because of the custom shaders I wrote for the above, but I tend to agree that a small quad shader seems like a good approach here, assuming we quash the NaNs.

I think I've got rid of them (at least in my iOS device) in version 2.0.2, tomorrow I'll test the screenshot generation again

Reply:
Thanks for sharing the URL! This is what I get on a M2 Pro:

```
Loading 4k images
Download 4k images time: 33ms
CPU Decode time for 4k image: 1488ms
Loading 8k images
Download 8k images time: 172ms
CPU Decode time for 4k image: 6241ms
```

> is there a better approach to a custom shader that would more easily translate between WebGL & WebGPU?

Looking at the decoding times, I would recommend to implement the parsing in a Web Worker. In this way, the main thread isn't blocked. If necessary, apps could bypass the parsing time with an intermediate environment map.

Reply:
Hmm, so this is the function that needs to be faster... 🤔

```js
for (let i = 0, l = sdr.data.length; i < l; i += originalChannels) {
	// these need to be converted to Linear from sRGB
	const r = SRGBToLinear(sdr.data[i + 0] / 255.0)
	const g = SRGBToLinear(sdr.data[i + 1] / 255.0)
	const b = SRGBToLinear(sdr.data[i + 2] / 255.0)

	// these are already in linear space
	const recoveryR = gainMap.data[i + 0] / 255.0
	const recoveryG = gainMap.data[i + 1] / 255.0
	const recoveryB = gainMap.data[i + 2] / 255.0

	const logRecoveryR = Math.pow(recoveryR, 1.0 / gamma[0])
	const logRecoveryG = Math.pow(recoveryG, 1.0 / gamma[1])
	const logRecoveryB = Math.pow(recoveryB, 1.0 / gamma[2])

	const logBoostR = gainMapMin[0] * (1.0 - logRecoveryR) + gainMapMax[0] * logRecoveryR
	const logBoostG = gainMapMin[1] * (1.0 - logRecoveryG) + gainMapMax[1] * logRecoveryG
	const logBoostB = gainMapMin[2] * (1.0 - logRecoveryB) + gainMapMax[2] * logRecoveryB

	const R = (r + offsetSdr[0]) * Math.pow(2, logBoostR * weightFactor) - offsetHdr[0]
	const G = (g + offsetSdr[1]) * Math.pow(2, logBoostG * weightFactor) - offsetHdr[1]
	const B = (b + offsetSdr[2]) * Math.pow(2, logBoostB * weightFactor) - offsetHdr[2]

	out[outIndex + 0] = DataUtils.toHalfFloat(Math.min(65504, Math.max(-65504, R)))
	out[outIndex + 1] = DataUtils.toHalfFloat(Math.min(65504, Math.max(-65504, G)))
	out[outIndex + 2] = DataUtils.toHalfFloat(Math.min(65504, Math.max(-65504, B)))
	out[outIndex + 3] = DataUtils.toHalfFloat(Math.min(65504, Math.max(-65504, 1)))

	outIndex += targetChannels
}
```

Reply:
> Hmm, so this is the loop that needs to be faster... 🤔


exactly. it is a simple line by line implementation of [the spec published by adobe](https://helpx.adobe.com/content/dam/help/en/camera-raw/using/gain-map/jcr_content/root/content/flex/items/position/position-par/table/row-io13dug-column-4a63daf/download_section/download-1/Gain_Map_1_0d14.pdf)  under **Pseudocode for Applying Gain Map**

I'm a bit stumped by your reasoning because, if you look closely at it even in their spec they seem to assume the "applying" is done in pseudo-shader code. It is in fact a very simple shader
```glsl

uniform sampler2D sdr;
uniform sampler2D gainMap;
uniform vec3 gamma;
uniform vec3 offsetHdr;
uniform vec3 offsetSdr;
uniform vec3 gainMapMin;
uniform vec3 gainMapMax;
uniform float weightFactor;

varying vec2 vUv;

void main() {
  vec3 rgb = texture2D(sdr, vUv).rgb;
  vec3 recovery = texture2D(gainMap, vUv).rgb;
  vec3 logRecovery = pow(recovery, gamma);
  vec3 logBoost = gainMapMin * (1.0 - logRecovery) + gainMapMax * logRecovery;
  vec3 hdrColor = (rgb + offsetSdr) * exp2(logBoost * weightFactor) - offsetHdr;
  gl_FragColor = vec4( hdrColor, 1.0 );
}
```

in its final "integrated" form, I imagine this would be done directly in your shaders (sort of like like with [RGBADepthPacking](https://github.com/mrdoob/three.js/blob/75cb81336da500afaad9b862ed86821e1bd9146f/src/renderers/shaders/ShaderChunk/packing.glsl.js#L24) which is able to represent float values using 8bit rgba) but I thought it would be simpler, for starters, to have it done separately and just use the renderTarget to populate other textures.

Well, anyway, I can re-introduce the JS "decoder" in the library but I must ask one final time if this is really the preferred method over the GPU implementation. It would be like re-implementing the `PMREMGenerator` in JS and let it calculate its values in JS instead of the current implementation.

I feel this is a similar situation.

Reply:
Yeah. I think we just needed more clarification about what the options were.

I think @elalish may be the best one to decide what path should be go with here.

Reply:
Good discussion, but I think this is a clear case where the shader is both simpler and much faster than JS. As a 3D library we shouldn't be afraid of shaders, and I'd certainly prefer a shader to a web worker. Now that the bug is fixed and this is working reliably across platforms, I'd say we should merge it. I certainly intend to adopt it.

Reply:
I was curious to see if that for loop could be optimised so I poked around for a bit...

Turns out it's the `SRGBToLinear` call that takes the most time.

With `SRGBToLinear`:

<img width="285" alt="Screenshot 2023-11-16 at 22 09 22" src="https://github.com/mrdoob/three.js/assets/97088/e493b404-1e69-403b-954c-233b38756afb">

Without `SRGBToLinear`:

<img width="281" alt="Screenshot 2023-11-16 at 22 09 36" src="https://github.com/mrdoob/three.js/assets/97088/1e5826f5-2d0d-4ec0-abb4-ee0cddb28104">